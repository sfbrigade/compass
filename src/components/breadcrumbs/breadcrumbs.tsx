import { trpc } from "@/client/lib/trpc";
import { useRouter } from "next/router";
import { SelectableForTable } from "zapatos/schema";

const Breadcrumbs = () => {
  const router = useRouter();
  const paths = router.asPath.split("/");

  let personData:
    | SelectableForTable<"student">
    | SelectableForTable<"user">
    | undefined;

  if (paths[1] === "students" && paths[2]) {
    const { data } = trpc.student.getStudentById.useQuery(
      { student_id: paths[2] },
      { enabled: Boolean(paths[2]), retry: false }
    );
    personData = data;
  } else if (paths[1] === "staff" && paths[2]) {
    const { data } = trpc.para.getParaById.useQuery(
      { user_id: paths[2] },
      { enabled: Boolean(paths[2]), retry: false }
    );
    personData = data;
  }

  return (
    <div
      style={{
        position: "fixed",
        display: "flex",
        left: "200px",
        zIndex: "1",
        padding: "1rem 0 1rem 2rem",
        width: "100vw",
        backgroundColor: "var(--grey-100)",
      }}
    >
      {paths.map((path, index) => {
        if (index === 0) return;
        if (index === 2)
          return (
            <div key={index}>
              / {personData?.first_name} {personData?.last_name}&nbsp;
            </div>
          );
        if (index % 2 === 1)
          return <div key={index}> / {path.toUpperCase()}&nbsp;</div>;
        else return <div key={index}> / {path}&nbsp;</div>;
      })}
    </div>
  );
};

export default Breadcrumbs;
