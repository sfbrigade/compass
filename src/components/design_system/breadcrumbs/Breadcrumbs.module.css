.container {
  position: fixed;
  display: flex;
  left: 200px;
  z-index: 1;
  padding: 1rem 0 1rem 2rem;
  width: 100%;
  background-color: var(--grey-100);
}

.link {
  text-decoration: none;
  color: var(--grey-10);
}

.link:hover {
  text-decoration: underline;
}
