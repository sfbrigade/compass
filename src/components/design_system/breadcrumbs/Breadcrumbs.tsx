/** DESIGN SYSTEM COMPONENT PLACEHOLDER
 * 1) Make a local branch for organizing your component (e.g. "design-systems-button")
 * 2) Replace this file and the corresponding css file(s) with your component file(s), cleaning up any duplicate files that are outside of the design components folder.
 * 3) Search and find all use cases for your component (likely linting will tell you where they are) and update the import paths
 * 4) Check code for errors and delete this comment
 * 5) Push code to branch and do a PR referencing the specific issue task you took for issue # 255.
 * NOTE: If you want a css.d.ts file to be generated to help with any type issues, you can run `npm run type-css`
 * */
export {};
