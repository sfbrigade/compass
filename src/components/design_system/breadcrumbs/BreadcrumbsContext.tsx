import {
  createContext,
  useCallback,
  useContext,
  useState,
  Dispatch,
  ReactNode,
  SetStateAction,
} from "react";

import type { Breadcrumb } from "./Breadcrumbs";

interface BreadcrumbsContextValue {
  breadcrumbs: Breadcrumb[] | undefined;
  setInitialBreadcrumbs: Dispatch<SetStateAction<Breadcrumb[] | undefined>>;
  setBreadcrumbs: (breadcrumbs: Breadcrumb[] | undefined) => void;
}

const breadcrumbsContext = createContext<BreadcrumbsContextValue>({
  breadcrumbs: undefined,
  setInitialBreadcrumbs: () => undefined,
  setBreadcrumbs: () => undefined,
});

export function useBreadcrumbsContext() {
  return useContext(breadcrumbsContext);
}

interface BreadcrumbsContextProviderProps {
  children: ReactNode;
}

export function BreadcrumbsContextProvider({
  children,
}: BreadcrumbsContextProviderProps) {
  const [breadcrumbs, setBreadcrumbs] = useState<Breadcrumb[] | undefined>();
  const setBreadcrumbsCallback = useCallback(
    (breadcrumbs: Breadcrumb[] | undefined) => {
      setTimeout(() => {
        setBreadcrumbs(breadcrumbs);
      }, 0);
    },
    [setBreadcrumbs]
  );
  return (
    <breadcrumbsContext.Provider
      value={{
        breadcrumbs,
        setInitialBreadcrumbs: setBreadcrumbs,
        setBreadcrumbs: setBreadcrumbsCallback,
      }}
    >
      {children}
    </breadcrumbsContext.Provider>
  );
}
