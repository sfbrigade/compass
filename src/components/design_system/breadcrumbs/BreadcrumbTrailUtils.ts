import { BreadcrumbContext } from "../../breadcrumbs/StatefulBreadcrumbTrailContext";

/**
 * Represents a single segment or "crumb" of a Breadcrumbs
 */
export interface BreadcrumbData {
  name: string;
  path: string;
  linkable: boolean;
}

export const transformPaths = ({
  fullPath,
  contextData,
}: {
  fullPath: string;
  contextData: BreadcrumbContext | undefined;
}): BreadcrumbData[] => {
  return fullPath
    .split("/")
    .map((path, index) => {
      // 0th index seems to only be empty string
      if (index === 0 || path === "") return null; // TODO Refactor so there is no first empty path element.

      // 1st index currently is either students or staff
      if (index % 2 === 1) {
        return {
          name: path.toUpperCase(),
          path: `/${path}`,
          linkable: true,
        } as BreadcrumbData;
      }

      // 2nd index is the ID referencing 1st index
      if (index === 2) {
        return {
          name: `${contextData?.person?.first_name ?? ""} ${
            contextData?.person?.last_name ?? ""
          }`,
          path: `/${path}`,
          linkable: false,
        } as BreadcrumbData;
      }
      return {
        name: path,
        path: `/${path}`,
        linkable: false,
      } as BreadcrumbData;
    })
    .filter((nullable) => nullable !== null) as BreadcrumbData[]; // TODO Remove filter when no longer needed.
};
