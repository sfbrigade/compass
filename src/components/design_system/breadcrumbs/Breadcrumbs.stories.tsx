import type { Meta, StoryObj } from "@storybook/react";

import Breadcrumbs from "./Breadcrumbs";
import type { Breadcrumb } from "./Breadcrumbs";

const meta: Meta<typeof Breadcrumbs> = {
  title: "Components/Design System/Breadcrumbs",
  component: Breadcrumbs,
};
export default meta;

type Story = StoryObj<typeof Breadcrumbs>;

const data: Breadcrumb[] = [
  {
    href: "/students",
    children: "Students",
  },
  {
    href: "/students/:student_id",
    children: "Student Name",
  },
  {
    href: "/students/:student_id/goals/:goal_id",
    children: "Goal #1",
  },
  {
    children: "Create Benchmark",
  },
];

export const Primary: Story = {
  args: {
    data,
  },
};
