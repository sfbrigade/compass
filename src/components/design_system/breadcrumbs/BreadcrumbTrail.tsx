import Breadcrumbs from "@mui/material/Breadcrumbs";
import $breadcrumbs from "./BreadcrumbTrail.module.css";
import { transformPaths } from "./BreadcrumbTrailUtils";
import { BreadcrumbContext } from "@/components/breadcrumbs/StatefulBreadcrumbTrailContext";
import Link from "next/link";

const BreadcrumbTrail = ({
  fullPath,
  contextData,
}: {
  fullPath: string;
  contextData: BreadcrumbContext | undefined;
}) => {
  // An array of breadcrumbs fixed to students/staff as the first index. This will be modified depending on how the address bar will be displayed.
  const breadcrumbs = transformPaths({ fullPath, contextData });

  return (
    <div className={$breadcrumbs.container}>
      <Breadcrumbs separator="/" aria-label="breadcrumb">
        {breadcrumbs.map((data, index) => {
          const isCurrentPage = index + 1 === breadcrumbs.length;

          if (!isCurrentPage && data.linkable) {
            return (
              <Link key={index} href={data.path} className={$breadcrumbs.link}>
                {data.name}
              </Link>
            );
          } else {
            return (
              <div key={index} className={$breadcrumbs.nonLinkCrumb}>
                {data.name}
              </div>
            );
          }
        })}
      </Breadcrumbs>
    </div>
  );
};

export default BreadcrumbTrail;
