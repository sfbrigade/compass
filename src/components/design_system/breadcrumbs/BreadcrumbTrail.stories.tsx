import type { Meta, StoryObj } from "@storybook/react";

import BreadcrumbTrail from "./BreadcrumbTrail";
import { Student } from "../../breadcrumbs/StatefulBreadcrumbTrailContext";

const meta = {
  component: BreadcrumbTrail,
} satisfies Meta<typeof BreadcrumbTrail>;

export default meta;

type Story = StoryObj<typeof meta>;

export const EmptyBreadcrumbTrail: Story = {
  args: {
    fullPath: "/",
    contextData: undefined,
  },
};

export const StudentPageBreadcrumbTrail: Story = {
  args: {
    fullPath: "/students/student-id",
    contextData: {
      person: {
        first_name: "Alia",
        last_name: "Atreides",
      } as Student,
    },
  },
};

export const GoalPageBreadcrumbTrail: Story = {
  args: {
    fullPath: "/students/student-id/goals/goal-id",
    contextData: {
      person: {
        first_name: "Alia",
        last_name: "Atreides",
      } as Student,
    },
  },
};
