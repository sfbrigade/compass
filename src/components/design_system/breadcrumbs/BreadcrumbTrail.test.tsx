import "@testing-library/jest-dom";
import { describe, expect, test } from "@jest/globals";
import { render } from "@testing-library/react";
import BreadcrumbTrail from "./BreadcrumbTrail";
import { Student } from "../../breadcrumbs/StatefulBreadcrumbTrailContext";

describe("BreadcrumbTrail", () => {
  test("renders empty for /", () => {
    const fullPath = "/";
    const contextData = undefined;

    const breadcrumbComponent = render(
      <BreadcrumbTrail fullPath={fullPath} contextData={contextData} />
    );

    expect(breadcrumbComponent).toMatchSnapshot();
  });

  test("renders empty for /students/student-id", () => {
    const fullPath = "/students/student-id";
    const contextData = {
      person: {
        first_name: "Alia",
        last_name: "Atreides",
      } as Student,
    };

    const breadcrumbComponent = render(
      <BreadcrumbTrail fullPath={fullPath} contextData={contextData} />
    );

    expect(breadcrumbComponent).toMatchSnapshot();
  });
});
