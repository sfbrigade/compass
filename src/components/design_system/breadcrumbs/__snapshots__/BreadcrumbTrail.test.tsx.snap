// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbTrail renders empty for / 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          />
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        />
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BreadcrumbTrail renders empty for /students/student-id 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="link"
                href="/students"
              >
                STUDENTS
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <div
                class="nonLinkCrumb"
              >
                Alia Atreides
              </div>
            </li>
          </ol>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="link"
              href="/students"
            >
              STUDENTS
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            /
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <div
              class="nonLinkCrumb"
            >
              Alia Atreides
            </div>
          </li>
        </ol>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
